cc/symmetries.h

```diff
@@ -74,157 +75,164 @@ inline Symmetry Inverse(Symmetry sym) {
   for (int j = 0; j < N; ++j) {
-    const auto* s = src + j * col_stride;
-    auto* d = dst + j * row_stride;
+    auto s = src + j * num_channels;
     for (int i = 0; i < N; ++i) {
-      memcpy(d, s, col_stride * sizeof(T));
-      d += col_stride;
+      dst = std::copy_n(s, num_channels, dst);
       s += row_stride;
     }
   }
 }
 
-template <typename T, int N, int num_channels>
-inline void FlipRot90(const T* src, T* dst) {
-  MG_CHECK(src != dst);
-  const int col_stride = num_channels;
-  const int row_stride = col_stride * N;
+template <int N, int num_channels, typename SrcIt, typename DstIt>
+inline void FlipRot90(SrcIt src, DstIt dst) {
+  MG_CHECK(dst != src);
+  const int row_stride = num_channels * N;
   for (int j = 0; j < N; ++j) {
-    const auto* s = src + (N - 1 - j) * row_stride;
-    auto* d = dst + j * row_stride;
+    auto s = src + (N - 1 - j) * row_stride;
     for (int i = 0; i < N; ++i) {
-      memcpy(d, s, col_stride * sizeof(T));
-      d += col_stride;
-      s += col_stride;
+      dst = std::copy_n(s, num_channels, dst);
+      s += num_channels;
     }
   }
 }
 
-template <typename T, int N, int num_channels>
-inline void FlipRot180(const T* src, T* dst) {
-  MG_CHECK(src != dst);
-  const int col_stride = num_channels;
-  const int row_stride = col_stride * N;
+template <int N, int num_channels, typename SrcIt, typename DstIt>
+inline void FlipRot180(SrcIt src, DstIt dst) {
+  MG_CHECK(dst != src);
+  const int row_stride = num_channels * N;
   for (int j = 0; j < N; ++j) {
-    const auto* s = src + (N - 1) * row_stride + (N - 1 - j) * col_stride;
-    auto* d = dst + j * row_stride;
+    auto s = src + (N - 1) * row_stride + (N - 1 - j) * num_channels;
     for (int i = 0; i < N; ++i) {
-      memcpy(d, s, col_stride * sizeof(T));
-      d += col_stride;
+      dst = std::copy_n(s, num_channels, dst);
       s -= row_stride;
     }
   }
 }
 
-template <typename T, int N, int num_channels>
-inline void FlipRot270(const T* src, T* dst) {
-  MG_CHECK(src != dst);
-  const int col_stride = num_channels;
-  const int row_stride = col_stride * N;
+template <int N, int num_channels, typename SrcIt, typename DstIt>
+inline void FlipRot270(SrcIt src, DstIt dst) {
+  MG_CHECK(dst != src);
+  const int row_stride = num_channels * N;
   for (int j = 0; j < N; ++j) {
-    const auto* s = src + j * row_stride + (N - 1) * col_stride;
-    auto* d = dst + j * row_stride;
+    auto s = src + j * row_stride + (N - 1) * num_channels;
     for (int i = 0; i < N; ++i) {
-      memcpy(d, s, col_stride * sizeof(T));
-      d += col_stride;
-      s -= col_stride;
+      dst = std::copy_n(s, num_channels, dst);
+      s -= num_channels;
     }
   }
 }
 
-template <typename T, int N, int num_channels>
-inline void ApplySymmetry(Symmetry sym, const T* src, T* dst) {
+template <int N, int num_channels, typename SrcIt, typename DstIt>
+inline void ApplySymmetry(Symmetry sym, SrcIt src, DstIt dst) {
```

Comment:
From https://en.cppreference.com/w/cpp/algorithm/copy: "In practice, implementations of std::copy avoid multiple assignments and use bulk copy functions such as std::memmove if the value type is TriviallyCopyable".

And https://stackoverflow.com/questions/4707012/is-it-better-to-use-stdmemcpy-or-stdcopy-in-terms-to-performance

Reply:
Sorry for not being clearer. I was asking about the SrcIt and DstIt template arguments, not std::copy. In the previous version of the code, the compiler was also able to unroll the loops and then vectorize the entire thing, not just the std::copy.

After looking more closely at the code, I think only the TensorRT iterators will likely be adversely affected because of the branch, so I'm less worried about this. Still, it would be nice to look at the disassembly to be sure.